controller/lib/core/flow_integrator.cpp

```diff
@@ -14,63 +14,78 @@ limitations under the License.
-// (equivalently, where the venturi response has a lower slope).  But we still
-// want to handle the low-flow case as best we can.
-static constexpr float FLOW_PID_KU = 0.20f;
-static constexpr float FLOW_PID_TU = 5;
-static constexpr VolumetricFlow FLOW_PID_OUTPUT_MAX = ml_per_sec(100);
-
-// TODO: Make PID parameters configurable.
-
-FlowIntegrator::FlowIntegrator()
-    : flow_correction_pid_(
-          /*kp=*/0.6f * FLOW_PID_KU,
-          /*ki=*/1.2f * FLOW_PID_KU / FLOW_PID_TU,
-          /*kd=*/3 * FLOW_PID_KU * FLOW_PID_TU / 40,
-          // ProportionalTerm::ON_ERROR and DifferentialTerm::ON_MEASUREMENT
-          // makes for your "standard" PID.  TODO: perhaps these shouldn't even
-          // be options.
-          ProportionalTerm::ON_ERROR,        //
-          DifferentialTerm::ON_MEASUREMENT,  //
-          -FLOW_PID_OUTPUT_MAX.ml_per_sec(), //
-          FLOW_PID_OUTPUT_MAX.ml_per_sec()) {}
+FlowIntegrator::FlowIntegrator() = default;
 
 void FlowIntegrator::AddFlow(Time now, VolumetricFlow uncorrected_flow) {
-  VolumetricFlow flow = uncorrected_flow + flow_correction_;
-  Duration delta = now - last_flow_measurement_time_;
-  if (delta >= VOLUME_INTEGRAL_INTERVAL) {
-    volume_ += delta * (last_flow_ + flow) / 2.0f;
+  VolumetricFlow flow = uncorrected_flow + FlowCorrection();
+
+  // First time?  In that case, don't need to adjust the flow.  Just record
+  // this data point.
+  if (last_flow_measurement_time_ == std::nullopt) {
+    last_note_time_ = now;
+    last_note_vol_ = ml(0);
+    last_flow_measurement_time_ = now;
+    last_flow_ = flow;
+    return;
+  }
+
+  Duration dt = now - *last_flow_measurement_time_;
+  if (dt >= VOLUME_INTEGRAL_INTERVAL) {
+    volume_ += dt * (last_flow_ + flow) / 2.0f;
     last_flow_measurement_time_ = now;
     last_flow_ = flow;
   }
 }
 
 void FlowIntegrator::NoteExpectedVolume(Volume v) {
-  flow_correction_ = ml_per_sec(flow_correction_pid_.Compute(
-      last_flow_measurement_time_, volume_.ml(), v.ml()));
+  // We need at least one call to AddFlow before we can do anything.
+  if (last_note_time_ == std::nullopt ||
+      last_flow_measurement_time_ == std::nullopt ||
+      *last_note_time_ >= *last_flow_measurement_time_) {
+    return;
+  }
+
+  // Compute the new flow correction as follows.
+  //
+  // - Compute the time delta between now (last_flow_measurement_time_) and the
+  //   last call to NoteExpectedVolume.
+  Duration dt = *last_flow_measurement_time_ - *last_note_time_;
+
+  // - Compute be what our volume would have been if we hadn't applied any
+  //   flow correction during the last breath.
+  Volume v_uncorrected = volume_ - flow_correction_ * dt;
+
+  // - Compute the difference in volume between now and the previous breath
+  //   we would have seen if we hadn't applied correction.
+  Volume dv = v_uncorrected - last_note_vol_;
+
+  // - Assume that dv will remain constant.  Then we can compute the expected
```

Comment:
dv is essentially "volume drift over one breath", right?

Reply:
Yes, but to be explicit, the trick that makes this algorithm work (and the reason an earlier attempt at a similar algorithm failed) is that it's not `new_corrected_volume - old_corrected_volume` but rather `new_uncorrected_volume - old_corrected_volume`, where `new_uncorrected_volume` is what we would have gotten if we'd turned off flow correction at the beginning of this breath that is now concluding.

This lets us calculate the volume drift irrespective of the flow correction we had been using!